src/bin/nydus-image/deduplicate.rs

```diff
@@ -0,0 +1,361 @@
+        ChunkTable::select(&conn).map_err(|e| anyhow::Error::new(e).into())
+    }
+
+    fn get_blobs(&self) -> Result<Vec<BlobTable>, Box<dyn Error + Send + Sync>> {
+        let conn = self.conn.lock().unwrap();
+        BlobTable::select(&conn).map_err(|e| anyhow::Error::new(e).into())
+    }
+}
+
+pub struct Deduplicate<D: Database + Send + Sync> {
+    sb: RafsSuper,
+    db: D,
+}
+impl Deduplicate<SqliteDatabase> {
+    pub fn new(bootstrap_path: &Path, config: Arc<ConfigV2>, db_path: &str) -> Result<Self> {
+        let (sb, _) = RafsSuper::load_from_file(bootstrap_path, config, false)?;
+        let db = SqliteDatabase::new(db_path)?;
+        Ok(Self { sb, db })
+    }
+
+    /// save metadata to database: chunk and blob info
+    pub fn save_metadata(&mut self, _mode: Option<&str>) -> anyhow::Result<Vec<Arc<BlobInfo>>> {
+        let tree = Tree::from_bootstrap(&self.sb, &mut ())
+            .context("failed to load bootstrap for deduplicate")?;
+
+        // Create blob table and chunk table
+        self.db
+            .create_chunk()
+            .map_err(|e| anyhow!("Failed to create chunk: {}", e))?;
+        self.db
+            .create_blob()
+            .map_err(|e| anyhow!("Failed to create blob: {}", e))?;
+
+        // Save blob info to blob table
+        let blob_infos = self.sb.superblock.get_blob_infos();
+        for blob in &blob_infos {
+            self.db
+                .insert_blob(&BlobTable {
+                    blob_id: blob.blob_id().to_string(),
+                    blob_size: blob.uncompressed_size(),
+                })
+                .map_err(|e| anyhow!("Failed to insert blob: {}", e))?;
+        }
+
+        // Save chunk info to chunk table
+        let pre = &mut |t: &Tree| -> Result<()> {
+            let node = t.lock_node();
+            for chunk in &node.chunks {
+                let index: u32 = chunk.inner.blob_index();
+                // Get blob id
+                let chunk_blob_id = blob_infos[index as usize].blob_id();
+                // Insert chunk into chunk table
+                self.db
+                    .insert_chunk(&ChunkTable {
+                        chunk_blob_id,
+                        chunk_digest: chunk.inner.id().to_string(),
+                        chunk_compressed_size: chunk.inner.compressed_size(),
+                        chunk_uncompressed_size: chunk.inner.uncompressed_size(),
+                        chunk_compressed_offset: chunk.inner.compressed_offset(),
+                        chunk_uncompressed_offset: chunk.inner.uncompressed_offset(),
+                    })
+                    .map_err(|e| anyhow!("Failed to insert chunk: {}", e))?;
+            }
+            Ok(())
+        };
+        tree.walk_dfs_pre(pre)?;
+
+        Ok(self.sb.superblock.get_blob_infos())
+    }
+}
+
+pub trait Table: Sync + Send + Sized + 'static {
+    // clear table
+    fn clear(conn: &Connection) -> Result<(), rusqlite::Error>;
```

Comment:
Why use sqlite-related error type `rusqlite::Error` here?

Reply:
Modified to a generic type.